Assets/Mirror/Editor/Weaver/Readers.cs

```diff
@@ -10,47 +10,93 @@ public static class Readers
-                { Weaver.colorType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadColor") },
-                { Weaver.color32Type.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadColor32") },
-                { Weaver.quaternionType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadQuaternion") },
-                { Weaver.rectType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadRect") },
-                { Weaver.planeType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadPlane") },
-                { Weaver.rayType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadRay") },
-                { Weaver.matrixType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadMatrix4x4") },
-                { Weaver.guidType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadGuid") },
-                { Weaver.gameObjectType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadGameObject") },
-                { Weaver.NetworkIdentityType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadNetworkIdentity") },
-                { Weaver.transformType.FullName, Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadTransform") },
-                { "System.Byte[]", Resolvers.ResolveMethod(networkReaderType, CurrentAssembly, "ReadBytesAndSize") },
-            };
+                RegisterReaders(td.Resolve());
+            }
         }
 
+        static void RegisterReaders(TypeDefinition td)
+        {
+            foreach (MethodDefinition md in td.Methods)
+            {
+                foreach (CustomAttribute ca in md.CustomAttributes)
+                {
+                    if (ca.AttributeType?.FullName == "Mirror.NetworkReaderAttribute")
+                    {
+                        RegisterReader(md);
+                        break;
+                    }
+                }
+            }
+
+            foreach (TypeDefinition embedded in td.NestedTypes)
+            {
+                RegisterReaders(embedded);
+            }
+        }
+
+        static void RegisterReader(MethodDefinition md)
+        {
+            TypeReference parameterType;
+
+            if (md.IsStatic)
+            {
+                if (md.Parameters.Count != 1)
+                {
+                    Weaver.Error($"Reader {md.FullName} must have exactly one NetworkReader parameter and no others");
+                    return;
+                }
+
+                parameterType = md.Parameters[0].ParameterType;
+            }
+            else
+            {
+                if (md.Parameters.Count != 0)
+                {
+                    Weaver.Error($"Reader {md.FullName} must have exactly one NetworkReader parameter and no others");
```

Comment:
Message here is wrong, should say something like `"Reader {md.FullName} must be static and take one NetworkReader as parameter."` (We dont want to guide people to add stuff into NetworkReader, so tell them to make it static)

Reply:
```suggestion
                    Weaver.Error($"Reader {md.FullName} must be static and take one NetworkReader as parameter.");
```